src/core/lib/iomgr/resource_quota.cc

```diff
@@ -430,39 +451,53 @@ static bool rq_reclaim(grpc_resource_quota* resource_quota, bool destructive) {
  * ru_slice: a slice implementation that is backed by a grpc_resource_user
  */
 
-typedef struct {
+static void ru_slice_ref(void* p);
+static void ru_slice_unref(void* p);
+static const grpc_slice_refcount_vtable ru_slice_vtable = {
+    ru_slice_ref, ru_slice_unref, grpc_slice_default_eq_impl,
+    grpc_slice_default_hash_impl};
+
+namespace {
+struct ru_slice_refcount {
+  ru_slice_refcount(grpc_resource_user* resource_user, size_t size)
+      : resource_user(resource_user), size(size) {
+    base.vtable = &ru_slice_vtable;
+    base.sub_refcount = &base;
+  }
+  ~ru_slice_refcount() {
+    grpc_resource_user_free(resource_user, size);
+  }
+
+  void Ref() { refs.Ref(); }
+  void Unref() {
+    if (!refs.Unref()) {
+      return;
+    }
+    this->~ru_slice_refcount();
+    gpr_free(this);
+  }
+
   grpc_slice_refcount base;
-  gpr_refcount refs;
+  grpc_core::RefCount refs;
   grpc_resource_user* resource_user;
   size_t size;
-} ru_slice_refcount;
+};
+}  // namespace
 
 static void ru_slice_ref(void* p) {
   ru_slice_refcount* rc = static_cast<ru_slice_refcount*>(p);
-  gpr_ref(&rc->refs);
+  rc->Ref();
 }
 
 static void ru_slice_unref(void* p) {
   ru_slice_refcount* rc = static_cast<ru_slice_refcount*>(p);
-  if (gpr_unref(&rc->refs)) {
-    grpc_resource_user_free(rc->resource_user, rc->size);
-    gpr_free(rc);
-  }
+  rc->Unref();
 }
 
-static const grpc_slice_refcount_vtable ru_slice_vtable = {
-    ru_slice_ref, ru_slice_unref, grpc_slice_default_eq_impl,
-    grpc_slice_default_hash_impl};
-
 static grpc_slice ru_slice_create(grpc_resource_user* resource_user,
                                   size_t size) {
-  ru_slice_refcount* rc = static_cast<ru_slice_refcount*>(
-      gpr_malloc(sizeof(ru_slice_refcount) + size));
-  rc->base.vtable = &ru_slice_vtable;
-  rc->base.sub_refcount = &rc->base;
-  gpr_ref_init(&rc->refs, 1);
-  rc->resource_user = resource_user;
-  rc->size = size;
+  ru_slice_refcount* rc = new (gpr_malloc(sizeof(ru_slice_refcount) + size))
```

Comment:
Why not just use `grpc_core::New<>` here?

Reply:
Please note that `malloc`'s size is `sizeof(ru_slice_refcount) + size` and not `sizeof(ru_slice_refcount)`, and as a result we cannot use `grpc_core::New`.